nixos/modules/services/web-apps/invidious.nix

```
@@ -7,29 +7,60 @@ let
 
   settingsFile = settingsFormat.generate "invidious-settings" cfg.settings;
 
+  finalHmacKeyFile =
+    if cfg.hmacKeyFile != null
+    then cfg.hmacKeyFile
+    else "/var/lib/invidious/hmac_key";
+  generateHmac = cfg.hmacKeyFile == null;
+
   serviceConfig = {
     systemd.services.invidious = {
       description = "Invidious (An alternative YouTube front-end)";
       wants = [ "network-online.target" ];
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
 
-      script =
-        let
-          jqFilter = "."
-            + lib.optionalString (cfg.database.host != null) "[0].db.password = \"'\"'\"$(cat ${lib.escapeShellArg cfg.database.passwordFile})\"'\"'\""
-            + " | .[0]"
-            + lib.optionalString (cfg.extraSettingsFile != null) " * .[1]";
-          jqFiles = [ settingsFile ] ++ lib.optional (cfg.extraSettingsFile != null) cfg.extraSettingsFile;
-        in
-        ''
-          export INVIDIOUS_CONFIG="$(${pkgs.jq}/bin/jq -s "${jqFilter}" ${lib.escapeShellArgs jqFiles})"
-          exec ${cfg.package}/bin/invidious
-        '';
+      preStart = lib.optionalString generateHmac ''
+        if [[ ! -e "${finalHmacKeyFile}" ]]; then
+          ${pkgs.pwgen}/bin/pwgen 20 1 > "${finalHmacKeyFile}"
+          chmod 0600 "${finalHmacKeyFile}"
+        fi
+      '';
+
+      script = ''
+        configParts=()
+      ''
+      # autogenerated hmac_key
+      + lib.optionalString generateHmac ''
+        configParts+=("$(${pkgs.jq}/bin/jq -R '{"hmac_key":.}' <"${finalHmacKeyFile}")")
+      ''
+      # generated settings file
+      + ''
+        configParts+=("$(cat "${settingsFile}")")
+      ''
+      # optional database password file
+      + lib.optionalString (cfg.database.host != null) ''
+        configParts+=("$(${pkgs.jq}/bin/jq -R '{"db":{"password":.}}' <"${cfg.database.passwordFile}")")
+      ''
+      # optional extra settings file
+      + lib.optionalString (cfg.extraSettingsFile != null) ''
+        configParts+=("$(cat "${cfg.extraSettingsFile})")
+      ''
+      # explicitly specified hmac key file
+      + lib.optionalString (cfg.hmacKeyFile != null) ''
+        configParts+=("$(cat "${cfg.hmacKeyFile})")
```

These lines seem to have mismatch `"`'s, and you can always avoid `cat` here, and use `lib.escapeShellArgs` to have correct escaping:

```suggestion
        configParts+=("$(<${lib.escapeShellArg cfg.hmacKeyFile})")
```

---

Good catch, the quoting is indeed wrong. I do need `cat` though, because `jq` needs the file contents, not the file names in the final merge step.

---

You might have missed it or don't know it, but in Bash, `$(<foo)` is the same as `$(cat foo)` while avoiding an extra `cat`! I believe the code should work as I proposed it :)

---

Yeah that'd work too. The original suggestion didn't include the `<` though, so only the file path would be included. I've changed it to `<` now to stop the `cat` abuse :)

---

Oh weird, I can definitely see a `<` in my suggestion!